Rewriting the whole gene.searchMany module to make it better comply with Angular's best practices, and make each part of it more loosely coupled with the other parts.

rzelayafavila opened the pull request:
This included getting rid of a lot of JS code and some templates too.

rzelayafavila commented:
P.S. - Since so much changed, I would recommend taking a look at the file diffs using the 'Split' view (on the upper-right area of the 'Files changed' tab).

cgreene commented:
Wow! A net reduction of at least 100 lines of javascript (especially considering the added comments)!

cgreene commented:
Can you launch a version of this so that reviewers can see it? I know that was helpful for @mhuyck when looking at @dongbohu's PR.

rzelayafavila commented:
@cgreene: Sure! I will post a url of it here in a sec.

rzelayafavila commented:
Ok, @cgreene @dongbohu @mhuyck: http://165.123.67.154:8000/#/gene_search is where it's at.

dongbohu commented on interface/src/app/gene/searchMany/searchMany.js:
```diff
@@ -36,226 +36,151 @@ angular.module('adage.gene.searchMany', [
-    },
-    size: function() {
-      return queries.length;
+    controller: ['$scope', function($scope) {
```
Here you can use either a controller or a link function to set variables in the directive's scope. The Angular book you have (**AngularJS: Up and Running**) has a good discussion on which one should be used under certain scenario (see **page 216**). Since you are not doing any inter-directive communication here, link function is probably a better choice.

dongbohu commented on interface/src/app/gene/searchMany/searchMany.js:
```diff
@@ -36,226 +36,151 @@ angular.module('adage.gene.searchMany', [
-    controller: ['$scope', '$rootScope', 'SearchResults',
-      function($scope, $rootScope, SearchResults) {
-        $scope.loadingSearchResults = false;
+    controller: ['$scope', 'Gene', function($scope, Gene) {
```
It is worth considering whether all directives in this file should use link functions instead of controllers. (I didn't realize the difference until I wrote my own directive recently.)

dongbohu commented on interface/src/app/gene/searchMany/searchMany.js:
```diff
@@ -36,226 +36,151 @@ angular.module('adage.gene.searchMany', [
-          }
-          SearchResults.search(qparams);
-        };
+        $scope.loadingSearchResults = true;
```
Trivial question: **loadingSearchResult** is a boolean flag. Can we give it a better name?

mhuyck replied:
Can you suggest something you think would be better, @dongbohu? I like that it is very descriptive, personally.

mhuyck replied:
Hah! What's wrong with `foobar`? :-)  I once had a classmate who liked to use the number `17` as a variable in his calculus problems.

On this particular variable name, I'd say that `loadingSearchResults` is a good description of state. A function that did that would be called something like `loadSearchResults`. A subtle distinction, perhaps, but the difference is very clear to me.

Something in between such as `resultsLoading` would be shorter but just as clear, in my opinion.

rzelayafavila replied:
Ok, I think I will keep this variable name for now, but keep an eye out and see if it is confusing to anyone else, and if so I will change it. I will also keep this in mind when naming more variables in the future.

dongbohu commented on interface/src/app/gene/searchMany/searchMany.js:
```diff
@@ -264,94 +189,80 @@ angular.module('adage.gene.searchMany', [
+      searchResults: '=',
+      queries: '='
+    },
     replace: true,
```
Not really related to this PR: The keyword **replace** has been deprecated since Angular 1.3. You may want to avoid it.

mhuyck replied:
Good to know. It appears this change is somewhat controversial: http://stackoverflow.com/questions/24194972/why-is-replace-deprecated-in-angularjs

dongbohu replied:
Good to know. I heard it was deprecated but never dug deeper.

mhuyck commented on interface/src/app/gene/gene-network.tpl.html:
```diff
@@ -3,7 +3,7 @@ <h1 class="page-header">
   </h1>
   <div class="well">
-    <gene-search-form></gene-search-form>
+    <gene-search-panel organism="organism"></gene-search-panel>
```
Interesting. I just noticed that the template corresponding to this directive is named `gene-search-form.tpl.html` rather than `geneSearchForm.tpl.html`. I didn't know that either style could work! The tutorials I used when learning Angular all had this `camelCase.tpl.html` template file name style mixed with `<camel-case>...` directive name style.

Since you have the hyphenated style of naming everywhere, I guess it works! Wondering if we should standardize on one approach vs. the other. What are your thoughts, @dongbohu and @rzelayafavila?

dongbohu replied:
As far as I know, Angular only cares about whether the file specified by **templateUrl** exists and can be loaded, it doesn't really enforce any filename rules. The suffix "**.tpl.html**" is actually enforced by ngBoilerplace in **build.config.js** (lines 31-32). So I don't think we need to stick to a certain format, unless you guys want to.

rzelayafavila replied:
Yep, I think (as @dongbohu said) you can specify the location & name of the template file in whatever format (but build.config.js will care about only including template files that end in 'tpl.html', as we have it currently configured). I think the only other thing that needs to follow the strict hyphens <-> camelCase connection is the name of the directive itself (camelCase) and the name of the element it will replace in the html template (hyphens). I am totally fine if we decide to standardize on a specific format for template names though.

mhuyck commented on interface/src/app/gene/searchMany/gene-search-form.tpl.html:
```diff
@@ -1,29 +1,11 @@
+    ng-disabled="loadingSearchResults">
+    Search <i class="fa fa-search"></i>
+  </button>
+</form>
```
I really like how this is now a tightly-focused "widget". Nice work!

mhuyck commented on interface/src/app/gene/searchMany/gene-search-form.tpl.html:
```diff
@@ -1,29 +1,11 @@
-  </div>
-</div>
+<form class="form">
+  <textarea name="genes" rows="7" ng-model="geneQueries"
```
Should this "magic number" 7 be a parameter that can be specified in the `<gene-search-form>` directive (but default to 7)? That way the `gene-search-panel`, which is responsible for layout of this widget, can specify another "size," if needed.

mhuyck commented on interface/src/app/gene/searchMany/gene-search-form.tpl.html:
```diff
@@ -1,29 +1,11 @@
-</div>
+<form class="form">
+  <textarea name="genes" rows="7" ng-model="geneQueries"
+    placeholder="Search genes by symbol or Entrez ID">
```
Similar to the above comment, I like to make the placeholder text a parameter that can be specified. Just a suggestion.

rzelayafavila replied:
True - I agree that both can/maybe should be a parameter. I will make an issue for this.

mhuyck commented on interface/src/app/gene/searchMany/gene-search-panel.tpl.html:
```diff
@@ -0,0 +1,17 @@
+      </search-result-table>
+    </div>
+  </div>
+</div>
```
I really like how this is now tightly focused solely on layout of these two widgets. We should strive for this clean separation of responsibilities wherever we can. Even the naming (`panel` here vs. `form`, above) is helpfully clear.

rzelayafavila replied:
Glad this makes it more clear!

mhuyck commented on interface/src/app/gene/searchMany/search-buttonset.tpl.html:
```diff
@@ -1,7 +1,18 @@
+  <button class="btn btn-default" ng-click="nextGenePage()"
+    ng-if="additionalButtonPages">
+      More <i class="fa fa-angle-double-right"></i>
+  </button>
```
I take it all of these buttons always appear together as a button set. If so, I think this strikes a great balance between reusable module and the "granularity" of re-use.

mhuyck commented on interface/src/app/gene/searchMany/searchMany.js:
```diff
@@ -36,226 +36,151 @@ angular.module('adage.gene.searchMany', [
+    controller: ['$scope', 'Gene', function($scope, Gene) {
+      $scope.errors = null;
+
+      $scope.searchGenes = function() {
```
Agree that this belongs in the `form` widget, not in the `panel` layout. Good separation of concerns.

mhuyck commented on interface/src/app/gene/searchMany/searchMany.js:
```diff
@@ -36,226 +36,151 @@ angular.module('adage.gene.searchMany', [
+        $scope.$watchCollection('searchResults', function() {
+          $scope.queries = Object.keys($scope.searchResults);
+          $scope.totalResults = $scope.queries.length;
+          $scope.resultsForPage = $scope.updatePageGenes();
```
I had forgotten about `$watchCollection`. This seems like a reasonable way to keep this section updated.

rzelayafavila replied:
I agree - I had totally forgotten about it until this PR, but it is a great functionality that Angular provides.

mhuyck commented:
I really like that Code Climate is keeping us DRY, but it sure doesn't seem to handle boilerplate-related repetition very well. Maybe there is a tuning parameter (mass?) we can adjust to make it a little less obsessive.

cgreene commented:
@mhuyck : agree that it might be customizable, but we might carefully consider this scholarly contribution before spending a lot of time dedicated to improving it:
https://xkcd.com/1205/

I usually just take a quick peek and say: codeclimate's nuts in this case.

rzelayafavila commented:
@dongbohu: Thank you - I am just checking these issues and making sure the code is as good as it can be for this PR (and create the appropriate issues for outstanding things) - I will probably merge pretty soon after I finish this.

dongbohu commented:
About @mhuyck's question on codeclimate, I actually customized a duplication threshold on python programs to 50, (which is part of the reason why codeclimate is not so paranoid on python programs), but didn't do it for JavaScript programs. I will add that parameter to make codeclimate less noisy.

rzelayafavila commented:
Ok, I think I addressed all the concerns and filed the pertaining issue! I decided not to move the code in the directives' ``controller`` function to the ``link`` function since I think this code interacts with the functionality of parent- and sibling- directives. I also removed the one directive that I would have done this code move for, geneResultButton. This was also the directive that included the ``replace: true`` line, so we don't need to worry about that anymore either. AND the code got even shorter!!
